src/main/java/org/wildfly/security/auth/server/SubjectUtil.java

```diff
@@ -0,0 +1,167 @@
+        PublicKeyCredential publicKeyCredential = publicCredentials.getCredential(PublicKeyCredential.class);
+        if (publicKeyCredential != null) {
+            subject.getPublicCredentials().add(publicKeyCredential.getPublicKey());
+        }
+
+        // TODO: should we process other public and private credentials?
+
+        return subject;
+    }
+
+    public static SecurityIdentity toSecurityIdentity(final SecurityDomain securityDomain, final Subject subject) {
+        Assert.checkNotNullParam("securityDomain", securityDomain);
+        Assert.checkNotNullParam("subject", subject);
+
+        return toSecurityIdentity(securityDomain, subject, securityDomain.getDefaultRealmName());
+    }
+
+    public static SecurityIdentity toSecurityIdentity(final SecurityDomain securityDomain, final Subject subject, final String realmName) {
+        Assert.checkNotNullParam("securityDomain", securityDomain);
+        Assert.checkNotNullParam("subject", subject);
+
+        // retrieve the principal name from the subject (the first non group principal found in the principals set).
+        Principal subjectPrincipal = null;
+        for (Principal principal : subject.getPrincipals()) {
+            if (!(principal instanceof Group)) {
+                subjectPrincipal = principal;
+                break;
+            }
+        }
+        // TODO: should we check for the 'CallerPrincipal' group and use that as the identity's principal?
+
+        // build an authorization identity using the subject groups as attributes.
+        Attributes attributes = null;
+        for (Principal principal : subject.getPrincipals()) {
+            if (principal instanceof Group) {
+                final String key = principal.getName();
+                final Set<String> values = new HashSet<>();
+                final Enumeration<? extends Principal> enumeration = ((Group) principal).members();
+                while (enumeration.hasMoreElements()) {
+                    values.add(enumeration.nextElement().getName());
+                }
+                if (attributes == null) {
+                    attributes = new MapAttributes();
+                }
+                attributes.addAll(key, values);
+            }
+        }
+        if (attributes == null)
+            attributes = Attributes.EMPTY;
+        final AuthorizationIdentity subjectAuthorizationIdentity = AuthorizationIdentity.basicIdentity(attributes);
+
+        // TODO: should we try to obtain public and private credentials from the subject?
+
+        // build the security identity using the supplied domain and the information extracted from the subject.
+        SecurityIdentity securityIdentity = new SecurityIdentity(securityDomain, subjectPrincipal,
+                securityDomain.getRealmInfo(realmName), subjectAuthorizationIdentity, securityDomain.getCategoryRoleMappers(),
+                SecurityIdentity.NO_PEER_IDENTITIES, IdentityCredentials.NONE, IdentityCredentials.NONE);
```

Comment:
Also I think the public/private credentials of the Subject should be forwarded on if possible.  But this would happen automatically as a result of using the existing authentication APIs.